### PR TITLE
cl_khr_fp16 support + LLVM 16 support

### PR DESCRIPTION
**Pull request by franz**

### CODE REVIEW

File: cmake/LLVM.cmake [hunk 1]
```diff
@@ -857,12 +857,35 @@ endif()
...
+#     For targets without F16C feature or above, please make sure:
+#     Use GCC 12.0 and above if you are using libgcc.
+#     If you are using compiler-rt, use the same version with the compiler.
+#        Early versions provided FP16 builtins in a different ABI. A workaround is
+#        to use a small code snippet to check the ABI if you cannot make sure of it.
+
+if(ENABLE_HOST_CPU_DEVICES AND NOT DEFINED CL_DISABLE_HALF)
+  # LLVM <15 doesn't support FP16 emulation
+  # LLVM 15 crashes on some code, with the FP16 emulation
+  # CONFORMANCE disables FP16 b/c it's incomplete
```
Comment (pjaaskel) [inline cmake/LLVM.cmake]: Can you document what is missing from conformance?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1600,7 +1600,10 @@ if (ENABLE_LEVEL0)
   pkg_check_modules(LEVEL0 REQUIRED level-zero>=1.3)
 
   if(NOT ENABLE_LLVM)
-    message(FATAL_ERROR "Level Zero requires LLVM/Clang")
+    message(FATAL_ERROR "Level Zero requires LLVM/Clang <= 15")
+  endif()
+  if(LLVM_VERSION VERSION_GREATER_EQUAL 16.0)
```
Comment (pjaaskel) [inline CMakeLists.txt:1616]: Can you add a comment that states what breaks currently (whatever you know about it)?

File: lib/CL/devices/devices.c [hunk 3]
```diff
@@ -514,7 +514,13 @@ pocl_init_devices ()
 #ifdef __linux__
 
 #ifdef ENABLE_HOST_CPU_DEVICES
+/* disabled by default if ENABLE_CONFORMANCE=OFF as it might
```
Comment (pjaaskel) [inline lib/CL/devices/devices.c:517]: Can you update the documentation of POCL_SIGFPE_HANDLER with this logic?

File: .github/workflows/build_cmake.yml [hunk 4]
```diff
@@ -49,8 +51,8 @@ jobs:
         id: cmake
         run: |
           runCMake() {
-            BUILD_FLAGS="-O1 -march=native -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-unused-variable"
-            cmake -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
+            BUILD_FLAGS="-O1 -march=native -Wall -Wextra -Wno-unused-parameter -Wno-unused-variable"
+            cmake -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DDEVELOPER_MODE=ON \
```
Comment (pjaaskel) [inline .github/workflows/build_cmake.yml:55]: It would be good to test at least some of the envs in non-devel mode.

File: examples/oneapi-samples/src/1d_HeatTransfer/src/1d_HeatTransfer.cpp [hunk 5]
```diff
@@ -0,0 +1,294 @@
+﻿//==============================================================
```
Comment (pjaaskel) [inline examples/oneapi-samples/src/1d_HeatTransfer/src/1d_HeatTransfer.cpp:1]: Could we avoid copying of the src via a git clone preparation step?

File: lib/CL/devices/level0/pocl-level0.cc [hunk 6]
```diff
@@ -1064,8 +1064,11 @@ int pocl_level0_alloc_mem_obj(cl_device_id ClDevice, cl_mem Mem, void *HostPtr)
     P->mem_ptr = Mem->mem_host_ptr;
     P->version = Mem->mem_host_ptr_version;
   } else {
-    //    void *all = Device->allocDeviceMem(mem->size);
-    Allocation = Device->allocSharedMem(Mem->size);
+    bool Compress = false;
+    if (pocl_get_bool_option("POCL_LEVEL0_COMPRESS", 0)) {
```
Comment (pjaaskel) [inline lib/CL/devices/level0/pocl-level0.cc:1068]: This env should be more descriptive (POCL_LEVEL0_COMPRESS_RO_BUFFERS or similar) and documented. Also, is there a reason not to use this in case compression is available? It's best to avoid extra code paths if possible.

**Review by pjaaskel — CHANGES_REQUESTED** [review]

**Comment by pjaaskel** [comment]
TCE fails to build:
```
[  1%] Building CL to LLVM bitcode /home/visit0r/src/pocl/test-pr-cmake-llvm-15-tce-tests-build/lib/kernel/tce/tta-le/cbrt.cl.bc
/home/visit0r/src/pocl/lib/kernel/cbrt.cl:27:1: error: use of undeclared identifier '__builtin_cbrtf16'
DEFINE_BUILTIN_V_V(cbrt)
^
/home/visit0r/src/pocl/lib/kernel/templates.h:109:12: note: expanded from macro 'DEFINE_BUILTIN_V_V'
    return __builtin_##NAME##f16(a);                    \
           ^
<scratch space>:10:1: note: expanded from here
__builtin_cbrtf16

```

File: examples/oneapi-samples/src/1d_HeatTransfer/src/1d_HeatTransfer.cpp (hunk 5 above)
Comment (franz) [reply examples/oneapi-samples/src/1d_HeatTransfer/src/1d_HeatTransfer.cpp:1]: Possibly later, but ATM it does not make sense. The oneapi-samples git repo has an unusable (for PoCL) buildsystem (Cmake files with hardcoded `icpx` compiler, sometimes not even CMake just plain handwritten Makefile, each sample directory is a separate CMake / Make project ... etc), so we have to use our own CMake files anyway. If the oneapi-samples  buildsystem is fixed someday, then we could use the repo directly.

Comment (pjaaskel) [reply examples/oneapi-samples/src/1d_HeatTransfer/src/1d_HeatTransfer.cpp:1]: Can you document the reasons for copying in the commit log?

File: lib/CL/devices/level0/pocl-level0.cc (hunk 6 above)
Comment (franz) [reply lib/CL/devices/level0/pocl-level0.cc:1068]: I don't want it documented (yet), this is an experimental flag. I want to measure if it has any performance impact at all (so far i haven't seen any, but haven't done much testing). But good point about the code paths, i'll just enable the compression if it proves useful and remove the variable.

File: CMakeLists.txt (hunk 2 above)
Comment (franz) [reply CMakeLists.txt:1616]: Clang 16 produces bitcode with opaque pointers (despite PoCL passing -no-opaque-pointers). SPIR-V translator seems to be Ok with it, but the Level0 driver crashes inside the compiler.  I'll add a comment.

File: cmake/LLVM.cmake (hunk 1 above)
Comment (franz) [reply cmake/LLVM.cmake]: fixed

File: .github/workflows/build_cmake.yml (hunk 4 above)
Comment (franz) [reply .github/workflows/build_cmake.yml:55]: fixed

File: CHANGES [hunk 7]
```diff
@@ -33,12 +34,18 @@ Notable User Facing Changes
...
+- Partial support for cl_khr_fp16 when compiled with LLVM 16+. The
+  implementation relies on Clang, and may result in emulation if
+  the CPU does not support the required instruction set.
 
 Notable Fixes
 -------------
 
 - Fix handling of clBuildProgram() -I directories given with quotes or
   with leading/trailing white space.
+- CPU devices now flush printf() output after each call instead of waiting
```
Comment (pjaaskel) [inline CHANGES:46]: Why did you move this to fixes section? I think it could be considered a feature strictly put.

File: lib/CL/devices/devices.c (hunk 3 above)
Comment (franz) [reply lib/CL/devices/devices.c]: added

File: CHANGES (hunk 7 above)
Comment (franz) [reply CHANGES]: fixed

**Comment by pjaaskel** [comment]
CUDA build issue:
```
/home/visit0r/src/pocl/lib/kernel/acos.cl:30:1: error: use of undeclared identifier '__builtin_acosf16'
DEFINE_BUILTIN_V_V(acos)
^
/home/visit0r/src/pocl/lib/kernel/templates.h:102:12: note: expanded from macro 'DEFINE_BUILTIN_V_V'
    return __builtin_##NAME##f16(a);                    \
           ^
<scratch space>:10:1: note: expanded from here
__builtin_acosf16
```
